File: Data Structures & Algorithms/two-integer-sum/submission-3.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // int n = nums.size();
        // for(int i =0; i<n ; i++){
        //     for(int j = i+1 ; j<n ; j++){
        //         if(nums[i] + nums[j] == target){
        //             return {i,j};
        //         }
        //     }
        // }

        int i = 0;
        int j = nums.size() - 1;

        while (i < j)
        {
        int sum = nums[i] + nums[j];

        if (sum == target)
            return {i, j};

        else if (sum > target)
            j--;

        else
            i++;
        }

       return {};
    }
};

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-3.py
class Solution:
    def subarraySum(self, nums, k):

        n = len(nums)
        count = 0

        prefixSum = [0] * n
        prefixSum[0] = nums[0]

        # Prefix Sum Array
        for i in range(1, n):
            prefixSum[i] = prefixSum[i - 1] + nums[i]

        mp = {}

        for j in range(n):

            if prefixSum[j] == k:
                count += 1

            val = prefixSum[j] - k

            if val in mp:
                count += mp[val]

            if prefixSum[j] not in mp:
                mp[prefixSum[j]] = 0

            mp[prefixSum[j]] += 1

        return count